controlplane/src/core/repositories/FederatedGraphRepository.ts

```diff
@@ -216,17 +235,105 @@ export class FederatedGraphRepository {
     });
   }
 
-  public async list(opts: ListFilterOptions): Promise<FederatedGraphDTO[]> {
+  public move(data: { targetId: string; newNamespace: string; updatedBy: string }, blobStorage: BlobStorage) {
+    return this.db.transaction(async (tx) => {
+      const namespaceRepo = new NamespaceRepository(tx, this.organizationId);
+      const fedGraphRepo = new FederatedGraphRepository(tx, this.organizationId);
+      const subgraphRepo = new SubgraphRepository(tx, this.organizationId);
+      const compositionRepo = new GraphCompositionRepository(tx);
+
+      const newNS = await namespaceRepo.byName(data.newNamespace);
+      if (!newNS) {
+        throw new PublicError(
+          EnumStatusCode.ERR_NOT_FOUND,
+          `Namespace ${data.newNamespace} not found. Please create it before moving.`,
+        );
+      }
+
+      const federatedGraph = await fedGraphRepo.byTargetId(data.targetId);
+      if (!federatedGraph) {
+        throw new PublicError(EnumStatusCode.ERR_NOT_FOUND, `Federated graph not found`);
+      }
+
+      await tx.update(targets).set({ namespaceId: newNS.id }).where(eq(targets.id, data.targetId));
+
+      // Delete all mappings because we will deal with new subgraphs in new namespace
+      await tx
+        .delete(schema.subgraphsToFederatedGraph)
+        .where(eq(schema.subgraphsToFederatedGraph.federatedGraphId, federatedGraph.id));
+
+      const newNamespaceSubgraphs = await subgraphRepo.byGraphLabelMatchers(federatedGraph.labelMatchers, newNS.name);
+
+      // insert new mappings
+      if (newNamespaceSubgraphs.length > 0) {
+        await tx
+          .insert(schema.subgraphsToFederatedGraph)
+          .values(
+            newNamespaceSubgraphs.map((sg) => ({
+              subgraphId: sg.id,
+              federatedGraphId: federatedGraph.id,
+            })),
+          )
+          .onConflictDoNothing()
+          .execute();
+      }
+
+      const composer = new Composer(fedGraphRepo, subgraphRepo, compositionRepo);
+      const composedGraph = await composer.composeFederatedGraph(federatedGraph);
+
+      await composer.deployComposition({
+        composedGraph,
+        composedBy: data.updatedBy,
+        blobStorage,
+        organizationId: this.organizationId,
+      });
+
+      return composedGraph.errors;
+    });
+  }
+
+  public async listAll(opts: Omit<ListFilterOptions, 'namespace'>): Promise<FederatedGraphDTO[]> {
     const targets = await this.db.query.targets.findMany({
       where: and(eq(schema.targets.type, 'federated'), eq(schema.targets.organizationId, this.organizationId)),
       limit: opts.limit,
       offset: opts.offset,
+      orderBy: asc(schema.targets.namespaceId),
+    });
```

Comment:
I'd prefer to write a single query that joins the federated graphs table instead of doing N queries for M graphs.

Reply:
We do it later